### PR TITLE
Rearrange BlobStoreRepository to Prepare #46250

### PR DESCRIPTION
**Pull request by original-brownbear**
In #46250 we will have to move to two different delete
paths for BwC. Both paths will share the initial
reads from the repository though so I extracted/separated the
delete logic away from the initial reads to significantly simplify
the diff against #46250 here.
Also, I added some JavaDoc from #46250 here as well which makes the
code a little easier to follow even ignoring #46250 I think.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -377,55 +377,66 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
...
+     * @param repositoryData    RepositoryData found the in the repository before executing this delete
+     * @param listener          Listener to invoke once finished
+     */
+    private void doDeleteShardSnapshots(SnapshotId snapshotId, long repositoryStateId, Map<String, BlobContainer> foundIndices,
+                                        Map<String, BlobMetaData> rootBlobs, RepositoryData repositoryData,
+                                        ActionListener<Void> listener) throws IOException {
+        final RepositoryData updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+        // Cache the indices that were found before writing out the new index-N blob so that a stuck master will never
+        // delete an index that was created by another master node after writing this index-N blob.
+        writeIndexGen(updatedRepositoryData, repositoryStateId);
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I might be missing something: why do we write the updated index gen here before `getSnapshotInfo(snapshotId);`? Also, could getSnapshotInfo(snapshotId); be based on the rootBlobs (at least to check if it is missing)?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Basically this is just for safety and simplicity. Do the fewest possible steps to get the deletion done first, then worry about cleaning up.
We could also load the `SnapshotInfo` first, but it wouldn't change anything, but lead to (albeit to a trivial degree) more failed deletes.

But yea, we could do that check based on the root blobs, the question is if that gets us much (saves one GET in the corner case of a corrupted repo, but that's it pretty much isn't it)?

Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Thanks :+1: I was worried to have missed something.



> But yea, we could do that check based on the root blobs, the question is if that gets us much (saves one GET in the corner case of a corrupted repo, but that's it pretty much isn't it)?

Yes, let's keep it as it is.

**Review by tlrx — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**